### PR TITLE
Ensure customers do not have to replicate stored proc parameters in the config.

### PR DESCRIPTION
**Pull request by rohkhann**
Stored procedure today, is trying to validate that number of params in config == number of params in sp definition. this makes it difficult for customer as they have to respecify all the parameters in the  sp in the config again. Our validation should only ever fail in two scenarios:
1. The customer provides inputs in the config that dont match whats there in the stored proc. 
2. The customer does not provide input in the query and there is no backing default in the config. (ideally if there was a default on the db stored proc we should have detected it but it seems very hard for sql). As a workaround, customer can just define that default in the config if they want. 

Difficulties in determining whether a stored procedure parameter has default values:
1. Documented by microsoft - 
image
https://learn.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-sql-modules-transact-sql?view=sql-server-ver16
2. There are some alternatives but they are flaky -https://copyprogramming.com/howto/get-default-value-of-stored-procedure-parameter 

We can try the SQL DOM approach but the focus on this pr is to ensure that irrespective of whether there is or is not a default parameter customers can query their endpoint. If there is no query parameter provided and nothing specified in config we error out (even if for now they have a default defined in the db.)

### CODE REVIEW

**Comment by aaronburtle** [comment]
Changes look good, but I did not see a negative case to provide test coverage for the cases you mention in the description about validation failing in the config validation tests or in the sp builder tests. Can you double check if we have a test of this, and if not add an issue to track adding if not adding to this PR itself?

**Review by aaronburtle — APPROVED** [review]

File: src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs [hunk 1]
```diff
@@ -37,24 +38,30 @@ public static class GraphQLStoredProcedureBuilder
             // which are needed because parameter and column names can differ.
             StoredProcedureDefinition spdef = (StoredProcedureDefinition)dbObject.SourceDefinition;
 
-            // Create input value definitions from parameters defined in runtime config.
-            if (entity.Source.Parameters is not null)
+            // Create input value definitions from parameters defined in stored proc.
+            if (spdef is not null)
             {
-                foreach (string param in entity.Source.Parameters.Keys)
+                foreach ((string param, ParameterDefinition definition) in spdef.Parameters)
```
Comment (seantleonard) [inline src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs:44]: So by iterating over the resolved Stored procedure definition retreived from the db, you can handle the case when the customer provides default values for the input param in the runtime config that dont match whats there in the stored proc?

**Review by seantleonard — CHANGES_REQUESTED** [review]
Need tests to validate scenario that you are trying to accommodate. 
- If input parameter in db schema has a default value and 1) none is supplied in request 2) no default value defined in runtime config -> what is the expected outcome? 

What kind of test do you think could validate your failure case proposal:
> The customer does not provide input in the query and there is no backing default in the config. (ideally if there was a default on the db stored proc we should have detected it but it seems very hard for sql). As a workaround, customer can just define that default in the config if they want.

**Review by Aniruddh25 — CHANGES_REQUESTED** [review]
Need validation of at least the params that are specified

File: src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs [hunk 2]
```diff
@@ -37,24 +38,30 @@ public static class GraphQLStoredProcedureBuilder
...
                 {
                     // Input parameters defined in the runtime config may denote values that may not cast
                     // to the exact value type defined in the database schema.
                     // e.g. Runtime config parameter value set as 1, while database schema denotes value type decimal.
                     // Without database metadata, there is no way to know to cast 1 to a decimal versus an integer.
-                    string defaultValueFromConfig = entity.Source.Parameters[param].ToString()!;
-                    Tuple<string, IValueNode> defaultGraphQLValue = ConvertValueToGraphQLType(defaultValueFromConfig, parameterDefinition: spdef.Parameters[param]);
+
+                    IValueNode? valueNode = null;
+                    if (entity.Source.Parameters is not null && entity.Source.Parameters.TryGetValue(param, out object? value))
```
[the remark this answers is not in the document]
Comment (ayush3797) [reply src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs:52]: Instead of retrieving the default value, is it not sufficient to just have the metadata which tells whether there is a default value available for a stored proc param? In that case, we can just invoke the SP without passing in a value for that param in case no value was specified by the user.

**Comment by ayush3797** [comment]
Ideally, I think these cases make sense to me:

1. The parameter does not have a default value specified in the stored proc definition and in the config:
-> User needs to provide value for that param
2. The parameter does not have a default value specified in the SP definition but has a default value specified in the config:
-> User does not need to provide value for the param and we will use the value from the config.
3. The parameter has a default value specified in the SP definition and in the config (do we want to allow this scenario?): 
-> User does not need to provide value for the param and we will use the value from the config.
4. The parameter has a default value specified in the SP definition but not in the config:
-> User does not need to provide value for the param and we will use the value from the SP definition. 

For doing this, we would need the metadata about whether a param has default value in the SP definition or not.

File: src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs (hunk 1 above)
Comment (rohkhann) [reply src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs:44]: This is already validated while creating the stored procedure definition here: 
image

File: src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohkhann) [reply src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs:52]: Ive added description in the pr header on why this is very hard to do. The existing solutions are flaky and it requires proper review. However, this pr main aim is to unblock you from having to readd every parameter from your sql db into your config even without lack of default values. Now atleast customer can respecify the default value in the config or have it in their frontend code and atleast make queries without having to repeat every parameter in the config.

File: src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs [hunk 3]
```diff
@@ -37,24 +38,30 @@ public static class GraphQLStoredProcedureBuilder
...
+
                     inputValues.Add(
                         new(
                             location: null,
                             name: new(param),
                             description: new StringValueNode($"parameters for {name.Value} stored-procedure"),
-                            type: new NamedTypeNode(defaultGraphQLValue.Item1),
-                            defaultValue: defaultGraphQLValue.Item2,
+                            type: new NamedTypeNode(SchemaConverter.GetGraphQLTypeFromSystemType(type: definition.SystemType)),
+                            defaultValue: valueNode,
```
[the remark this answers is not in the document]
Comment (rohkhann) [reply src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs:64]: config vs sp validation is already done at time of creation of the stored proc. When we reach here every parameter in the config is validated to exist in the stored proc: 
image

File: src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs [hunk 4]
```diff
@@ -37,24 +38,30 @@ public static class GraphQLStoredProcedureBuilder
...
+                foreach ((string param, ParameterDefinition definition) in spdef.Parameters)
                 {
                     // Input parameters defined in the runtime config may denote values that may not cast
                     // to the exact value type defined in the database schema.
                     // e.g. Runtime config parameter value set as 1, while database schema denotes value type decimal.
                     // Without database metadata, there is no way to know to cast 1 to a decimal versus an integer.
-                    string defaultValueFromConfig = entity.Source.Parameters[param].ToString()!;
-                    Tuple<string, IValueNode> defaultGraphQLValue = ConvertValueToGraphQLType(defaultValueFromConfig, parameterDefinition: spdef.Parameters[param]);
+
+                    IValueNode? valueNode = null;
```
[the remark this answers is not in the document]
Comment (rohkhann) [reply src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs]: fixed.

**Comment by Aniruddh25** [comment]
> Ideally, I think these cases make sense to me:

1. The parameter does not have a default value specified in the stored proc definition and in the config:
-> User needs to provide value for that param
2. The parameter does not have a default value specified in the SP definition but has a default value specified in the config:
-> User does not need to provide value for the param and we will use the value from the config.
3. The parameter has a default value specified in the SP definition and in the config (do we want to allow this scenario?): 
-> User does not need to provide value for the param and we will use the value from the config.
4. The parameter has a default value specified in the SP definition but not in the config:
-> User does not need to provide value for the param and we will use the value from the SP definition. 

For doing this, we would need the metadata about whether a param has default value in the SP definition or not. 

Looks like the problem is SP definition cannot reliably provide us information whether the parameter has a default or not + we dont know what the value would be.

**Comment by Aniruddh25** [comment]
LGTM. Thanks for fixing this! Would be nice if we can have a follow up for checking SP metadata using sys.sql_modules or OBJECT_DEFINITION as the docs suggest for finding if the parameter has default or not.

**Review by Aniruddh25 — APPROVED** [review]

File: src/Service.Tests/SqlTests/GraphQLQueryTests/MsSqlGraphQLQueryTests.cs [hunk 5]
```diff
@@ -374,6 +375,22 @@ public async Task QueryAgainstSPWithOnlyTypenameInSelectionSet()
             await QueryAgainstSPWithOnlyTypenameInSelectionSet(dbQuery);
         }
 
+        /// <summary>
+        /// Checks failure on providing arguments with no default in runtimeconfig.
+        /// </summary>
```
Comment (seantleonard) [inline src/Service.Tests/SqlTests/GraphQLQueryTests/MsSqlGraphQLQueryTests.cs:380]: Can you elaborate on the state of the system under test in the comment:
1. Which fields are expected to be present in the request for the request to pass? 
2. What runtime config properties are included/excluded to ensure this test is in the proper state? e.g. no default values specified in runtime config 
3. What are the relevant db schema traits of the stored procedure under test? e.g. defines X number of default value params which are named a, b, and c.

I think one more test is warranted where the GraphQL request includes the a value for the field, where the field:
- has default value defined in db
- no default value defined in config

Result: successful response.

Comment (rohkhann) [reply src/Service.Tests/SqlTests/GraphQLQueryTests/MsSqlGraphQLQueryTests.cs:380]: This test seems to be already confirmed by this test case right: 
image

Comment (seantleonard) [reply src/Service.Tests/SqlTests/GraphQLQueryTests/MsSqlGraphQLQueryTests.cs]: The test in the screenshot may have unintentionally covered the use case I mentioned. 

If that pre-existing test has its underlying entity swapped out, then covering this test case may no longer apply, even though the test in the screenshot still accomplishes what it claims to be testing :

> The GraphQL request includes a value for the field, where the field:
> 
> - has default value defined in db
> - no default value defined in config

-> Action: Update the comment on that test `TestStoredProcedureQueryForGettingSingleRow(string dbQuery)` to elaborate on the state of the system under test like in my first comment of this thread.

File: src/Service.Tests/Snapshots/ConfigurationTests.TestReadingRuntimeConfigForMsSql.verified.txt [hunk 6]
```diff
@@ -2095,16 +2095,13 @@
...
-            id: 1
-          }
+          Type: stored-procedure
         },
         GraphQL: {
           Singular: GetPublisher,
           Plural: GetPublishers,
           Enabled: true,
-          Operation: Mutation
+          Operation: Query
```
Comment (abhishekkumams) [inline src/Service.Tests/Snapshots/ConfigurationTests.TestReadingRuntimeConfigForMsSql.verified.txt:2104]: why this change?

**Review by seantleonard — APPROVED** [review]
Pending response to Abhishek's comment.

File: src/Service.Tests/Snapshots/ConfigurationTests.TestReadingRuntimeConfigForMsSql.verified.txt (hunk 6 above)
Comment (rohkhann) [reply src/Service.Tests/Snapshots/ConfigurationTests.TestReadingRuntimeConfigForMsSql.verified.txt:2104]: GetPublisher is a getter method returning results and hence should be marked as query as typically mutations are used to create/update/delete from db.

File: src/Service.Tests/SqlTests/GraphQLQueryTests/MsSqlGraphQLQueryTests.cs (hunk 5 above)
Comment (rohkhann) [reply src/Service.Tests/SqlTests/GraphQLQueryTests/MsSqlGraphQLQueryTests.cs]: Makes sense. added comment.
